### PR TITLE
Wait up to 2 seconds for connection to become available

### LINKED ISSUE #755: Discovery failed error is shown after network reconnect
1. Open app
2. disable wifi
3. enable wifi
4. Discovery failed error is shown, even when a local url ist set

2.0.19-beta
Android 7.1.2

### PR DESCRIPTION
**Pull request by mueller-ma**
Especially for local connections, like WiFi or maybe VPN, too, a
connection may not be "available" (~ connected to the internet) once it is
"connected". To allow a connection to do whatever is needed to do, wait up
to 2 seconds (in 10 200ms chunks) for the connection to become available.
If it is not available after this timeframe, the connection is returned as
"not reachable".

Fixes #755
Closes #780

Signed-off-by: Florian <florian.schmidt.welzow@t-online.de>
Signed-off-by: mueller-ma <mueller-ma@users.noreply.github.com>

### CODE REVIEW

File: mobile/src/main/java/org/openhab/habdroid/core/connection/AbstractConnection.java [hunk 1]
```diff
@@ -84,14 +86,38 @@ public boolean checkReachabilityInBackground() {
...
+            Log.d(TAG, e.getMessage());
             return false;
         }
     }
+
+    private Socket createConnectedSocket(InetSocketAddress socketAddress) {
+        Socket s = new Socket();
+        for (int retries = 0; retries < 10; retries++) {
+            try {
+                s.connect(socketAddress, 1000);
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/core/connection/AbstractConnection.java]: https://github.com/openhab/openhab-android/pull/780/files#r178436704 still applies: could return s here and remove the `continue` / `return s` below.

File: mobile/src/main/java/org/openhab/habdroid/core/connection/AbstractConnection.java [hunk 2]
```diff
@@ -84,14 +86,38 @@ public boolean checkReachabilityInBackground() {
...
+            } catch (SocketTimeoutException e) {
+                Log.d(TAG, "Socket timeout at the " + retries + ". try.", e);
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/core/connection/AbstractConnection.java]: `5. try` is not valid English, as `5.` is a German construct. Could write `Socket timeout after 5 retries` here.

Also, I'm not sure whether printing the stack trace here (and below) is actually useful.

File: mobile/src/main/java/org/openhab/habdroid/core/connection/AbstractConnection.java [hunk 3]
```diff
@@ -84,14 +86,38 @@ public boolean checkReachabilityInBackground() {
...
+                return null;
+            } catch (IOException e) {
+                Log.d(TAG, "Socket connection failed at the " + retries + ". try.", e);
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/core/connection/AbstractConnection.java]: As above, but here one could write `Socket creation failed (attempt 5)`

File: mobile/src/main/java/org/openhab/habdroid/core/connection/AbstractConnection.java (hunk 2 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/core/connection/AbstractConnection.java]: > Also, I'm not sure whether printing the stack trace here (and below) is actually useful.

No, I don't think so.

File: mobile/src/main/java/org/openhab/habdroid/core/connection/AbstractConnection.java (hunk 1 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/core/connection/AbstractConnection.java]: Done

File: mobile/src/main/java/org/openhab/habdroid/core/connection/AbstractConnection.java (hunk 3 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/core/connection/AbstractConnection.java]: Done

**Comment by maniac103** [comment]
@mueller-ma Forgot to push the update? ;-)

**Comment by mueller-ma** [comment]
Yes. Pushed it now.

File: mobile/src/main/java/org/openhab/habdroid/core/connection/AbstractConnection.java [hunk 4]
```diff
@@ -84,14 +86,38 @@ public boolean checkReachabilityInBackground() {
             } else if (url.getProtocol().equals("https") && checkPort == -1) {
                 checkPort = 443;
             }
-            Socket s = new Socket();
-            s.connect(new InetSocketAddress(url.getHost(), checkPort), 1000);
+            Socket s = createConnectedSocket(new InetSocketAddress(url.getHost(), checkPort));
+            if (s == null) {
+                return false;
+            }
             Log.d(TAG, "Socket connected");
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/core/connection/AbstractConnection.java]: Can drop this message now that `createConnectedSocket` generates virtually the same message.

Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/core/connection/AbstractConnection.java]: Done.

**Pull request merged**